crates/deltalake-core/src/operations/constraints.rs

```diff
@@ -0,0 +1,263 @@
+
+            let files_to_check =
+                find_files(&this.snapshot, this.log_store.clone(), &state, None).await?;
+            let scan = DeltaScanBuilder::new(&this.snapshot, this.log_store.clone(), &state)
+                .with_files(&files_to_check.candidates)
+                .build()
+                .await?;
+
+            let task_ctx = Arc::new(TaskContext::from(&state));
+            let record_stream: SendableRecordBatchStream = scan.execute(0, task_ctx)?;
+            let records = collect_sendable_stream(record_stream).await?;
+
+            for batch in records {
+                checker.check_batch(&batch).await?;
+            }
+
+            // We have validated the table passes it's constraints, now to add the constraint to
+            // the table.
+
+            metadata
+                .configuration
+                .insert(format!("delta.constraints.{}", name), Some(expr.clone()));
+
+            let old_protocol = this.snapshot.protocol();
+            let protocol = Protocol {
+                min_reader_version: if old_protocol.min_reader_version > 1 {
+                    old_protocol.min_reader_version
+                } else {
+                    1
+                },
+                min_writer_version: if old_protocol.min_reader_version > 3 {
+                    old_protocol.min_reader_version
+                } else {
+                    3
+                },
+                reader_features: old_protocol.reader_features.clone(),
+                writer_features: old_protocol.writer_features.clone(),
+            };
+
+            let operational_parameters = HashMap::from_iter([
+                ("name".to_string(), json!(&name)),
+                ("expr".to_string(), json!(&expr)),
+            ]);
+            let commit_info = CommitInfo {
+                timestamp: Some(Utc::now().timestamp_millis()),
+                operation: Some("ADD CONSTRAINT".to_string()),
+                operation_parameters: Some(operational_parameters),
+                read_version: Some(this.snapshot.version()),
+                isolation_level: Some(IsolationLevel::Serializable),
+                is_blind_append: Some(false),
+                ..Default::default()
+            };
+
+            let actions = vec![
+                Action::CommitInfo(commit_info),
+                Action::Metadata(Metadata::try_from(metadata)?),
+                Action::Protocol(protocol),
+            ];
+
+            let operations = DeltaOperation::AddConstraint {
+                name: name.clone(),
+                expr: expr.clone(),
+            };
+
+            let _version = commit(
+                this.log_store.as_ref(),
+                &actions,
+                operations,
+                &this.snapshot,
+                None,
+            )
+            .await?;
+
+            Ok(DeltaTable::new_with_state(this.log_store, this.snapshot))
```

Comment:
I don't think, this snapshot will reflect he latest table state including the updates we just did, as the `commit` function will not update the snapshot.

Reply:
So the correct thing here @roeap is to do a merge on the states after the commit right?

Reply:
Exactly, we should add the actions we just committed to the table also to the new state.